File: chapters/04_react_and_tool_stream/v2_structured_reasoning.ts
// TypeScript mirror for Chapter 04: ReAct and Tool Stream (v2)
// Run with: npx ts-node v2_structured_reasoning.ts

function reasonSteps(problem: string): string[] {
  const steps = ["Inspect error type", "Identify missing package"];
  if (problem.includes("requests")) {
    steps.push("Suggest installation command");
  } else {
    steps.push("Suggest dependency check");
  }
  return steps;
}

function reasonAndAnswer(problem: string): [string[], string] {
  const steps = reasonSteps(problem);
  const final = problem.includes("requests")
    ? "Run: pip install requests"
    : "Verify project dependencies and reinstall.";
  return [steps, final];
}

function main(): void {
  const [steps, final] = reasonAndAnswer("ModuleNotFoundError: No module named 'requests'");
  console.log(steps);
  console.log(final);
}

main();

File: chapters/04_react_and_tool_stream/v2_structured_reasoning.py
from __future__ import annotations


def reason_steps(problem: str) -> list[str]:
    steps = ["Inspect error type", "Identify missing package"]
    if "requests" in problem:
        steps.append("Suggest installation command")
    else:
        steps.append("Suggest dependency check")
    return steps


def reason_and_answer(problem: str) -> tuple[list[str], str]:
    steps = reason_steps(problem)
    if "requests" in problem:
        final = "Run: pip install requests"
    else:
        final = "Verify project dependencies and reinstall."
    return steps, final


def main() -> None:
    steps, final = reason_and_answer("ModuleNotFoundError: No module named 'requests'")
    print(steps)
    print(final)


if __name__ == "__main__":
    main()
